monorepo-migration/convert-to-hg.sh

```diff
@@ -0,0 +1,88 @@
+#!/bin/sh
+
+set -e
+set -x
+
+CINNABAR=https://github.com/jcristau/git-cinnabar
+CINNABAR_BRANCH=detect-more-renames
+SOURCE_REPO=https://github.com/mozilla-mobile/firefox-android
+SOURCE_REPO_NAME=${SOURCE_REPO##*/}
+SOURCE_BRANCH=main
+DEST_REPO="$(readlink -f "$1")"
+DEST_SUBDIR=mobile/android
+EXPRESSIONS_FILE_PATH="$(dirname $(readlink -f "$0"))/data/message-expressions.txt"
+
+TEMPDIR=$(mktemp -d)
+#trap 'rm -rf ${TEMPDIR} EXIT'
+CINNABAR_PATH="${TEMPDIR}/git-cinnabar"
+SOURCE_REPO_PATH="${TEMPDIR}/${SOURCE_REPO_NAME}"
+
+setup_temporary_repo() {
+  git clone "${SOURCE_REPO}" "${SOURCE_REPO_PATH}"
+  git -C "${SOURCE_REPO_PATH}" config user.email=release@mozilla.com --config user.name="Mozilla Release Engineering"
+}
+
+setup_cinnabar() {
+  git clone --branch "${CINNABAR_BRANCH}" --depth 1 "${CINNABAR}" "${CINNABAR_PATH}"
+  make -C "${CINNABAR_PATH}"
+}
+
+rewrite_git_history() {
+  # replace octopus merges with a series of two-parent merges
+  FILTER_BRANCH_SQUELCH_WARNING=1 \
+  git -C "${SOURCE_REPO_PATH}" filter-branch \
+    --force \
+    --commit-filter \
+      'if test "$#" -lt 7; then
+         git commit-tree "$@"
+       else
+         commit_message=$(mktemp)
+         cat > $commit_message
+             
+         final_tree="$1"
+         shift
+         shift
+         p1="$1"
+         shift
+         part=1
+         while test "$#" -ge 2; do
+           shift
+           p2="$1"
+           shift
+           tree=$(git merge-tree --write-tree --allow-unrelated-histories "$p1" "$p2")
+           p1="$(sed "1s/$/ (part $part)/" $commit_message | git commit-tree "$tree" -p "$p1" -p "$p2")"
+           part=$((part + 1))
+         done
+         rm -f $commit_message
+         git diff-tree --exit-code $tree $final_tree
+         echo "$p1"
+     fi' \
+        -- "$SOURCE_BRANCH"
+  # to-subdirectory-filter: move everything to $DEST_SUBDIR
+  # replace-message: replace references to pull requests with the full URL
+  # preserve-commit-hashes: because the filter-branch step also changes
+  #                         hashes, as does the conversion to hg, rewriting
+  #                         hashes here would only point at temporary commit
+  #                         objects which wouldn't help
```

Comment:
Question: Is `preserve-commit-hashes` faster than the default behavior? If not, I'm not sure I follow this comment 😅

Reply:
preserve-commit-hashes means we don't rewrite hashes in commit messages to refer to temporary commits that won't be living outside the mid-rewrite git repository, and they'll keep referring to pre-migration commit hashes in the firefox-android repo; that seems preferable to me.

Reply:
Okay. Will commit hashes still be present in commit messages? Would you have an example of the final result?

Reply:
Yes, for example https://github.com/mozilla-mobile/firefox-android/commit/5620f848cfc8e82ffa7fb22a402eeeeb945cccb1 becomes something like https://hg.mozilla.org/users/jcristau_mozilla.com/firefox-android-hg/rev/d3adbd8f0c94 and the reference to 3c660fe68b872a9f20618846a96528ab20bfefcf is left unchanged.

Reply:
Understood! I agree it's better to keep the original hash here rather than referring to one that will only exist on the machine handling the migration. However, I think it's going to confuse future readers. For instance, future readers may believe the referenced commit above is an HG rev, meaning https://hg.mozilla.org/users/jcristau_mozilla.com/firefox-android-hg/rev/3c660fe68b872a9f20618846a96528ab20bfefcf should be valid.

I think we can fix that by pre-prepending `https://github.com/mozilla-mobile/firefox-android/commit/` to the referenced commit. This seems to be feasible with `git filter-repo --commit-callback`[1]. Let me know what you think! 

[1] https://github.com/newren/git-filter-repo/blob/4bc9022afce5e2e138596bbecf4df310212ae164/Documentation/git-filter-repo.txt#L1178

Reply:
I'll take a stab at that, maybe in a followup PR?